core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java

```diff
@@ -94,21 +105,30 @@ protected void initChannel(Channel ch) throws Exception {
         p.addLast(new FlushConsolidationHandler());
         p.addLast(ReadSuppressingHandler.INSTANCE);
 
-        if (port.protocol().isTls()) {
-            assert sslContexts != null;
-            p.addLast(new SniHandler(sslContexts));
-            p.addLast(TrafficLoggingHandler.SERVER);
-            configureHttps(p);
+        if (config.acceptProxyProtocol()) {
+            p.addLast(new HAProxyMessageDecoder(config.proxyProtocolMaxTlvSize()));
+            p.addLast(new HAProxyMessageConsumer());
```

Comment:
Could we auto-detect between HTTPS, PROXY and HTTP instead of accepting only PROXY protocol?

Reply:
Perhaps we could add `SessionProtocol.PROXY`? e.g.

- If `ServerPort` contains `PROXY` and `HTTP`, support HTTP and proxied HTTP.
- If `ServerPort` contains `PROXY` and `HTTPS`, support HTTPS and proxied HTTPS.
- If `ServerPort` contains `PROXY`, `HTTP` and `HTTPS`, support HTTP, HTTPS, proxied HTTP and proxied HTTPS.
- If `ServerPort` contains `PROXY` only, `ServerBuilder.build()` fails.

Then we could replace `allowProxyProtocol` with port configuration methods such as:

```java
sb.http(443).https(443).haproxy(443);
```

Reply:
Good suggestion! I will change my code with `SessionProtocol.PROXY`. :-)